File: avalonia/NezAvalonia/Core/GamepadServer.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.NetworkInformation;
using System.Net.Sockets;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace NezAvalonia.Core;

/// <summary>
/// HTTP + WebSocket server that serves the web gamepad HTML page,
/// and receives controller input via WebSocket.
/// Mirror mode uses the incremental delta protocol over WebSocket.
/// </summary>
public sealed class GamepadServer : IDisposable
{
    private INezEngine? _engine;
    private DeltaEncoder _deltaEncoder;
    private HttpListener? _listener;
    private CancellationTokenSource? _cts;
    private Task? _listenerTask;
    private readonly ConcurrentBag<WebSocket> _sockets = new();

    // Track which sockets want mirror (delta) frames
    private readonly ConcurrentDictionary<WebSocket, bool> _mirrorClients = new();

    public int Port { get; private set; } = 8080;
    public bool IsRunning => _listener?.IsListening == true;
    public INezEngine? Engine => _engine;

    /// Create without an engine — engine will be set when gameplay starts.
    public GamepadServer() : this(null!) { }

    public GamepadServer(INezEngine engine)
    {
        _engine = engine;
        if (_engine != null) {
            _deltaEncoder = new DeltaEncoder(_engine.ScreenWidth, _engine.ScreenHeight, blockSize: 8);
        } else {
            // Placeholder: will be re-initialized when engine is set
            _deltaEncoder = new DeltaEncoder(256, 240, blockSize: 8);
        }
    }

    /// Update the active game engine (called when entering/exiting gameplay).
    public void SetEngine(INezEngine? engine)
    {
        _engine = engine;
        if (_engine != null)
            _deltaEncoder = new DeltaEncoder(_engine.ScreenWidth, _engine.ScreenHeight, blockSize: 8);
    }

    /// <summary>
    /// Get the local network IP address for QR code generation.
    /// </summary>
    public static string GetLocalIp()
    {
        // Method 1: UDP socket trick (most reliable cross-platform)
        try
        {
            using var socket = new System.Net.Sockets.Socket(
                AddressFamily.InterNetwork, System.Net.Sockets.SocketType.Dgram, 0);
            socket.Connect("8.8.8.8", 80);
            if (socket.LocalEndPoint is System.Net.IPEndPoint ep
                && ep.Address.ToString() != "127.0.0.1"
                && ep.Address.ToString() != "0.0.0.0")
                return ep.Address.ToString();
        }
        catch { }

        // Method 2: DNS hostname resolution
        try
        {
            var host = System.Net.Dns.GetHostEntry(System.Net.Dns.GetHostName());
            foreach (var addr in host.AddressList)
            {
                if (addr.AddressFamily == AddressFamily.InterNetwork
                    && !IPAddress.IsLoopback(addr)
                    && !addr.ToString().StartsWith("169.254"))
                    return addr.ToString();
            }
        }
        catch { }

        // Method 3: Enumerate network interfaces
        try
        {
            foreach (var ni in NetworkInterface.GetAllNetworkInterfaces())
            {
                if (ni.OperationalStatus != OperationalStatus.Up) continue;
                if (ni.NetworkInterfaceType is NetworkInterfaceType.Loopback
                    or NetworkInterfaceType.Tunnel) continue;

                foreach (var addr in ni.GetIPProperties().UnicastAddresses)
                {
                    if (addr.Address.AddressFamily == AddressFamily.InterNetwork
                        && !IPAddress.IsLoopback(addr.Address)
                        && !addr.Address.ToString().StartsWith("169.254")
                        && addr.Address.ToString() != "127.0.0.1")
                    {
                        return addr.Address.ToString();
                    }
                }
            }
        }
        catch { }

#if ANDROID
        // Method 4: Android WifiManager
        try
        {
            var context = Android.App.Application.Context;
            var wifiManager = (Android.Net.Wifi.WifiManager?)context.GetSystemService(Android.Content.Context.WifiService);
            if (wifiManager?.ConnectionInfo?.IpAddress is int ip && ip != 0)
            {
                return $"{ip & 0xFF}.{(ip >> 8) & 0xFF}.{(ip >> 16) & 0xFF}.{(ip >> 24) & 0xFF}";
            }
        }
        catch { }
#endif

        return "0.0.0.0"; // Signal that no IP was found
    }

    public void Start()
    {
        if (_listener != null) return;

        _cts = new CancellationTokenSource();
        _listener = new HttpListener();

        // Try multiple prefix styles for cross-platform compat
        try
        {
            _listener.Prefixes.Add($"http://*:{Port}/");
            _listener.Start();
        }
        catch
        {
            // macOS without admin rights: fall back to localhost + 0.0.0.0 via separate listener
            _listener = new HttpListener();
            try
            {
                _listener.Prefixes.Add($"http://+:{Port}/");
                _listener.Start();
            }
            catch
            {
                _listener = new HttpListener();
                _listener.Prefixes.Add($"http://localhost:{Port}/");
                _listener.Start();
            }
        }

        _listenerTask = Task.Run(() => AcceptLoop(_cts.Token));
    }

    public void Stop()
    {
        _cts?.Cancel();
        _deltaCts?.Cancel();
        try { _listener?.Stop(); } catch { /* ignore */ }
        _listener?.Close();
        _listener = null;

        // Close all websockets
        foreach (var ws in _sockets)
        {
            try
            {
                if (ws.State == WebSocketState.Open)
                    ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "Server stopping",
                        CancellationToken.None).Wait(500);
            }
            catch { /* ignore */ }
            ws.Dispose();
        }
        _sockets.Clear();
        _mirrorClients.Clear();
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested && _listener?.IsListening == true)
        {
            try
            {
                var ctx = await _listener.GetContextAsync().ConfigureAwait(false);
                _ = Task.Run(() => HandleRequest(ctx, ct), ct);
            }
            catch (ObjectDisposedException) { break; }
            catch (HttpListenerException) { break; }
            catch { /* continue */ }
        }
    }

    private async Task HandleRequest(HttpListenerContext ctx, CancellationToken ct)
    {
        try
        {
            var path = ctx.Request.Url?.AbsolutePath ?? "/";

            if (ctx.Request.IsWebSocketRequest && path == "/ws")
            {
                await HandleWebSocket(ctx, ct);
                return;
            }

            switch (path)
            {
                case "/":
                    ServeHtml(ctx);
                    break;
                default:
                    ctx.Response.StatusCode = 404;
                    ctx.Response.Close();
                    break;
            }
        }
        catch
        {
            try { ctx.Response.Close(); } catch { /* ignore */ }
        }
    }

    private void ServeHtml(HttpListenerContext ctx)
    {
        var html = Encoding.UTF8.GetBytes(WebGamepadHtml.Html);
        ctx.Response.ContentType = "text/html; charset=utf-8";
        ctx.Response.ContentLength64 = html.Length;
        ctx.Response.OutputStream.Write(html, 0, html.Length);
        ctx.Response.Close();
    }

    private async Task HandleWebSocket(HttpListenerContext ctx, CancellationToken ct)
    {
        WebSocketContext wsCtx;
        try
        {
            wsCtx = await ctx.AcceptWebSocketAsync(null);
        }
        catch
        {
            ctx.Response.StatusCode = 500;
            ctx.Response.Close();
            return;
        }

        var ws = wsCtx.WebSocket;
        _sockets.Add(ws);

        // Start delta broadcast if this is the first mirror client
        EnsureDeltaBroadcast(ct);

        var buf = new byte[1024];
        try
        {
            while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
            {
                var result = await ws.ReceiveAsync(new ArraySegment<byte>(buf), ct);
                if (result.MessageType == WebSocketMessageType.Close) break;
                if (result.MessageType != WebSocketMessageType.Text) continue;

                var json = Encoding.UTF8.GetString(buf, 0, result.Count);
                ProcessMessage(ws, json);
            }
        }
        catch { /* disconnected */ }
        finally
        {
            _mirrorClients.TryRemove(ws, out _);
            try
            {
                if (ws.State == WebSocketState.Open)
                    await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
            }
            catch { /* ignore */ }
            ws.Dispose();
        }
    }

    // ---- Delta frame broadcast ----

    private Task? _deltaBroadcastTask;
    private CancellationTokenSource? _deltaCts;

    private void EnsureDeltaBroadcast(CancellationToken ct)
    {
        // Already running
        if (_deltaBroadcastTask != null && !_deltaBroadcastTask.IsCompleted) return;

        _deltaCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        _deltaBroadcastTask = Task.Run(() => DeltaBroadcastLoop(_deltaCts.Token), _deltaCts.Token);
    }

    private async Task DeltaBroadcastLoop(CancellationToken ct)
    {
        var sw = Stopwatch.StartNew();
        double nextFrame = 0;
        const double frameInterval = 1000.0 / 30.0; // 30 fps delta stream

        try
        {
            while (!ct.IsCancellationRequested)
            {
                double now = sw.Elapsed.TotalMilliseconds;
                if (now < nextFrame)
                {
                    await Task.Delay(Math.Max(1, (int)(nextFrame - now)), ct);
                }
                nextFrame = Math.Max(nextFrame + frameInterval, sw.Elapsed.TotalMilliseconds);

                // Only broadcast if there are mirror clients
                if (_mirrorClients.IsEmpty) continue;

                var buffer = _engine.GetBackBuffer();
                if (buffer == null) continue;

                // Encode delta
                byte[] payload;
                try
                {
                    payload = _deltaEncoder.Encode(buffer);
                }
                catch
                {
                    continue;
                }

                // Send to all mirror clients
                var segment = new ArraySegment<byte>(payload);
                var deadSockets = new List<WebSocket>();

                foreach (var kvp in _mirrorClients)
                {
                    var ws = kvp.Key;
                    try
                    {
                        if (ws.State == WebSocketState.Open)
                            await ws.SendAsync(segment, WebSocketMessageType.Binary, true, ct);
                        else
                            deadSockets.Add(ws);
                    }
                    catch
                    {
                        deadSockets.Add(ws);
                    }
                }

                // Cleanup dead sockets
                foreach (var ws in deadSockets)
                {
                    _mirrorClients.TryRemove(ws, out _);
                }
            }
        }
        catch (OperationCanceledException) { }
        catch { /* ignore */ }
    }

    private void ProcessMessage(WebSocket ws, string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            var type = root.GetProperty("type").GetString();

            if (type == "btn")
            {
                int player = root.GetProperty("player").GetInt32();
                int btn = root.GetProperty("btn").GetInt32();
                bool pressed = root.GetProperty("pressed").GetBoolean();

                if (player == 1)
                    _engine.SetButton(btn, pressed);
                else if (player == 2)
                    _engine.SetButtonP2(btn, pressed);
            }
            else if (type == "turbo")
            {
                int player = root.GetProperty("player").GetInt32();
                string? btn = root.GetProperty("btn").GetString();
                bool active = root.GetProperty("active").GetBoolean();

                // Turbo only for P1 for now (matching existing engine API)
                if (player == 1)
                {
                    if (btn == "a") _engine.SetTurboA(active);
                    else if (btn == "b") _engine.SetTurboB(active);
                }
            }
            else if (type == "mirror")
            {
                bool active = root.GetProperty("active").GetBoolean();
                if (active)
                {
                    _mirrorClients[ws] = true;
                    // Force a full frame for new mirror client
                    _deltaEncoder.Reset();
                }
                else
                {
                    _mirrorClients.TryRemove(ws, out _);
                }
            }
            else if (type == "keyframe")
            {
                // Client requests a full keyframe (e.g. after reconnect or frame skip)
                _deltaEncoder.Reset();
            }
        }
        catch
        {
            // Ignore malformed messages
        }
    }

    public void Dispose()
    {
        Stop();
        _cts?.Dispose();
    }
}

File: flutter/lib/core/gamepad_server.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:flutter/foundation.dart';

import 'delta_encoder.dart';
import 'nez_engine.dart';
import 'web_gamepad_html.dart';

/// HTTP + WebSocket server for remote gamepad control and frame streaming.
/// Mirror mode uses the incremental delta protocol over WebSocket binary frames.
class GamepadServer {
  HttpServer? _server;
  final NezEngine _engine;
  final DeltaEncoder _deltaEncoder;
  final List<_WsClient> _clients = [];
  Timer? _frameTimer;
  String? _localIp;

  static const int port = 8080;
  static const int _frameIntervalMs = 33; // ~30fps delta stream

  GamepadServer(this._engine)
      : _deltaEncoder = DeltaEncoder(
          width: _engine.screenWidth,
          height: _engine.screenHeight,
          blockSize: 8,
          pixelFormat: DeltaEncoder.pixelFormatRgba32,
        );

  bool get isRunning => _server != null;
  String? get localIp => _localIp;
  String get p1Url => 'http://$_localIp:$port/?player=1';
  String get p2Url => 'http://$_localIp:$port/?player=2';
  String get p1MirrorUrl => 'http://$_localIp:$port/?player=1&mirror=true';
  String get p2MirrorUrl => 'http://$_localIp:$port/?player=2&mirror=true';

  Future<void> start() async {
    if (_server != null) return;

    _localIp = await _getLocalIp();
    if (_localIp == null) {
      debugPrint('NEZ GamepadServer: Could not determine local IP');
      _localIp = '127.0.0.1';
    }

    try {
      _server = await HttpServer.bind(InternetAddress.anyIPv4, port);
      debugPrint('NEZ GamepadServer: listening on $_localIp:$port');

      _server!.listen(_handleRequest, onError: (e) {
        debugPrint('NEZ GamepadServer error: $e');
      });

      // Push frames to mirror clients via WebSocket binary
      _frameTimer = Timer.periodic(
        const Duration(milliseconds: _frameIntervalMs),
        (_) => _pushFrames(),
      );
    } catch (e) {
      debugPrint('NEZ GamepadServer: failed to start: $e');
      _server = null;
    }
  }

  Future<void> stop() async {
    _frameTimer?.cancel();
    _frameTimer = null;

    for (final c in _clients) {
      c.ws.close();
    }
    _clients.clear();

    await _server?.close(force: true);
    _server = null;
    debugPrint('NEZ GamepadServer: stopped');
  }

  void _handleRequest(HttpRequest request) async {
    final path = request.uri.path;

    if (path == '/ws') {
      _handleWebSocket(request);
    } else {
      // Serve HTML for everything else
      request.response
        ..statusCode = 200
        ..headers.contentType = ContentType.html
        ..write(webGamepadHtml)
        ..close();
    }
  }

  void _handleWebSocket(HttpRequest request) async {
    try {
      final ws = await WebSocketTransformer.upgrade(request);
      final client = _WsClient(ws, wantsMirror: false);
      _clients.add(client);
      debugPrint('NEZ GamepadServer: WS client connected (${_clients.length})');

      ws.listen(
        (data) {
          if (data is String) {
            _processMessage(data, client);
          }
        },
        onDone: () {
          _clients.remove(client);
          debugPrint('NEZ GamepadServer: WS client disconnected');
        },
        onError: (e) {
          _clients.remove(client);
        },
      );
    } catch (e) {
      debugPrint('NEZ GamepadServer: WS upgrade failed: $e');
      request.response
        ..statusCode = 500
        ..close();
    }
  }

  void _processMessage(String data, _WsClient client) {
    try {
      final msg = jsonDecode(data) as Map<String, dynamic>;
      final type = msg['type'] as String?;
      final player = msg['player'] as int? ?? 1;

      if (type == 'btn') {
        final btn = msg['btn'] as int;
        final pressed = msg['pressed'] as bool;
        if (player == 1) {
          _engine.setButton(btn, pressed);
        } else {
          _engine.setButtonP2(btn, pressed);
        }
      } else if (type == 'turbo') {
        final btnName = msg['btn'] as String;
        final active = msg['active'] as bool;
        if (player == 1) {
          if (btnName == 'a') {
            _engine.setTurboA(active);
            _engine.setButton(0, active); // Also press A
          }
          if (btnName == 'b') {
            _engine.setTurboB(active);
            _engine.setButton(1, active); // Also press B
          }
        } else {
          // P2: just press the button (no turbo support yet)
          if (btnName == 'a') _engine.setButtonP2(0, active);
          if (btnName == 'b') _engine.setButtonP2(1, active);
        }
      } else if (type == 'reset') {
        _engine.togglePause();
        Future.delayed(const Duration(milliseconds: 100), () {
          _engine.togglePause();
        });
      } else if (type == 'mirror') {
        // Client requests frame streaming
        client.wantsMirror = msg['active'] as bool? ?? true;
        if (client.wantsMirror) {
          // Force a full keyframe for new mirror client
          _deltaEncoder.reset();
        }
        debugPrint('NEZ GamepadServer: client mirror=${client.wantsMirror}');
      } else if (type == 'keyframe') {
        // Client requests a full keyframe (e.g. after reconnect)
        _deltaEncoder.reset();
      }
    } catch (e) {
      debugPrint('NEZ GamepadServer: bad message: $e');
    }
  }

  void _pushFrames() {
    final hasMirrorClients = _clients.any((c) => c.wantsMirror);
    if (!hasMirrorClients) return;

    final rgba = _engine.lastRgbaFrame;
    if (rgba == null) return;

    // Encode delta (synchronous, fast — no JPEG encoding)
    Uint8List payload;
    try {
      payload = _deltaEncoder.encode(rgba);
    } catch (_) {
      return;
    }

    final toRemove = <_WsClient>[];
    for (final c in _clients) {
      if (!c.wantsMirror) continue;
      try {
        c.ws.add(payload);
      } catch (_) {
        toRemove.add(c);
      }
    }
    for (final c in toRemove) {
      _clients.remove(c);
      try { c.ws.close(); } catch (_) {}
    }
  }

  static Future<String?> _getLocalIp() async {
    try {
      final interfaces = await NetworkInterface.list(
        type: InternetAddressType.IPv4,
        includeLoopback: false,
      );
      // Prefer WiFi interfaces (wlan, en0, etc.)
      for (final iface in interfaces) {
        final name = iface.name.toLowerCase();
        if (name.contains('wlan') || name.contains('en0') || name.contains('wifi')) {
          for (final addr in iface.addresses) {
            if (!addr.isLoopback && !addr.address.startsWith('169.254')) {
              return addr.address;
            }
          }
        }
      }
      // Fallback: any non-loopback IPv4
      for (final iface in interfaces) {
        for (final addr in iface.addresses) {
          if (!addr.isLoopback && !addr.address.startsWith('169.254')) {
            return addr.address;
          }
        }
      }
    } catch (e) {
      debugPrint('NEZ GamepadServer: IP detection error: $e');
    }

    // Last resort: UDP socket trick
    try {
      final socket = await RawDatagramSocket.bind(InternetAddress.anyIPv4, 0);
      socket.close();
      // Try connecting to get local address
      final s = await Socket.connect('8.8.8.8', 80, timeout: const Duration(seconds: 2));
      final ip = s.address.address;
      s.destroy();
      if (ip != '127.0.0.1' && ip != '0.0.0.0') return ip;
    } catch (_) {}

    return null;
  }
}

class _WsClient {
  final WebSocket ws;
  bool wantsMirror;
  _WsClient(this.ws, {required this.wantsMirror});
}
